CRM/Event/BAO/Event.php

```diff
@@ -278,9 +278,7 @@ public static function getEvents(
       $endDate = date('YmdHis');
       $query .= "
         AND ( `end_date` >= {$endDate} OR
-          (
-            ( end_date IS NULL OR end_date = '' ) AND start_date >= {$endDate}
-          )
+          end_date IS NULL AND start_date >= {$endDate}
```

Comment:
Without the brackets it does happen to evaluate this as if it was `end_date >= {$endDate} OR (end_date IS NULL AND start_date >= {$endDate})`, but it might be better to explicitly add the brackets. (You can see this by typing `SELECT 1 OR 1 AND 0;`)

Otherwise this looks good. Note mariadb seems fine with it before even with the strict options, but agree the comparison to `''` doesn't seem needed.

Reply:
Oh hold on I just found where the comparison to `''` comes up. If you have strict settings off, and you insert `''` into a nullable datetime column, you get the `0000-00-00 00:00:00` value. The comparison to `''` finds those. And one additional quirk is that even if you enable strict mode after, it's perfectly happy keeping any existing 0000-00-00 values and treating the same as without strict in where clauses.

So we do need either the `''` or a comparison to `'0000-00-00'` - does mysql accept the latter in strict mode? (mariadb does)

Reply:
This will have to be `STR_TO_DATE('', '%Y%m%d%H%i%s')` (while the actual format doesn't really matter for `''`) then, as comparing date with string does not seem to be allowed in strict mode with MySQL 8.

Reply:
Oh clever.
When strict mode is on this doesn't find any lingering '0000-00-00' ones though, whereas before it did. But if someone had strict mode off, had some 0000's, and then turned strict mode on, then they kind of have invalid data at that point. So maybe the best way to deal with that is a status check that tells them to change those to nulls if they have strict mode on. (It would have to skip that check for mysql 8 though since this situation sounds like it can't exist in mysql 8, and such a check would error.)

We should add a code comment in here since someone looking at this code for the first time will be confused. I can do a followup.

Reply:
> It would have to skip that check for mysql 8 though since this situation sounds like it can't exist in mysql 8, and such a check would error.

As said, I was not able to find an `sql_mode` configuration in MySQL 8 that allowed the old query in any combination of `NO_ZERO_DATES`, `NO_ZERO_IN_DATES` and `ALLOW_INVALID_DATES`, and I'm not sure where "invalid" records with `0000-00-00 00:00:00` dates would result from. I think such dates should be converted to `NULL` anyway, so the system check might just check for invalid dates in the `civicrm_event` table in that follow-up.

Oh, and maybe it could correct the excess whitespace I accidentially added in the SQL before the first single opening `(` :roll_eyes:

Reply:
Yep: https://github.com/civicrm/civicrm-core/pull/29921